test/validation/api/timer/timer.c

```diff
@@ -641,6 +645,270 @@ static void timer_pool_max_res(void)
+	odp_pool_param_t pool_param;
+	odp_timer_start_t start_param;
+	odp_timer_pool_t tp;
+	odp_timer_t timer;
+	odp_pool_t pool;
+	odp_queue_t queue;
+	odp_timeout_t tmo;
+	odp_event_t ev;
+	odp_time_t t1, t2;
+	uint64_t tick, nsec;
+	int ret;
+	uint64_t tmo_ns = ODP_TIME_SEC_IN_NS;
+	uint64_t res_ns = ODP_TIME_SEC_IN_NS / 10;
+
+	memset(&capa, 0, sizeof(capa));
+	ret = odp_timer_capability(ODP_CLOCK_DEFAULT, &capa);
+	CU_ASSERT_FATAL(ret == 0);
+	CU_ASSERT_FATAL(capa.max_tmo.max_tmo > 0);
+
+	if (capa.max_tmo.max_tmo < tmo_ns) {
+		tmo_ns = capa.max_tmo.max_tmo;
+		res_ns = capa.max_tmo.res_ns;
+	}
+
+	odp_pool_param_init(&pool_param);
+	pool_param.type    = ODP_POOL_TIMEOUT;
+	pool_param.tmo.num = 10;
+	pool = odp_pool_create("timeout_pool", &pool_param);
+	CU_ASSERT_FATAL(pool != ODP_POOL_INVALID);
+
+	odp_queue_param_init(&queue_param);
+	queue_param.type = queue_type;
+
+	if (queue_type == ODP_QUEUE_TYPE_SCHED)
+		queue_param.sched.sync = ODP_SCHED_SYNC_ATOMIC;
+
+	queue = odp_queue_create("timer_queue", &queue_param);
+	CU_ASSERT_FATAL(queue != ODP_QUEUE_INVALID);
+
+	odp_timer_pool_param_init(&tp_param);
+
+	tp_param.res_ns     = res_ns;
+	tp_param.min_tmo    = tmo_ns / 4;
+	tp_param.max_tmo    = tmo_ns;
+	tp_param.num_timers = 1;
+
+	tp = odp_timer_pool_create("test_single", &tp_param);
+	CU_ASSERT_FATAL(tp != ODP_TIMER_POOL_INVALID);
+
+	odp_timer_pool_start();
+
+	timer = odp_timer_alloc(tp, queue, USER_PTR);
+	CU_ASSERT_FATAL(timer != ODP_TIMER_INVALID);
+
+	tmo = odp_timeout_alloc(pool);
+	ev  = odp_timeout_to_event(tmo);
+	CU_ASSERT_FATAL(ev != ODP_EVENT_INVALID);
+
+	nsec = tmo_ns;
+	if (restart)
+		nsec = tmo_ns / 2;
+
+	tick = odp_timer_ns_to_tick(tp, nsec);
+	if (tick_type == ODP_TIMER_TICK_ABS)
+		tick += odp_timer_current_tick(tp);
+
+	start_param.tick_type = tick_type;
+	start_param.tick      = tick;
+	start_param.tmo_ev    = ev;
+
+	ret = odp_timer_start(timer, &start_param);
+	CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
+
+	if (restart) {
+		tick = odp_timer_ns_to_tick(tp, tmo_ns);
+		if (tick_type == ODP_TIMER_TICK_ABS)
+			tick += odp_timer_current_tick(tp);
+
+		start_param.tick   = tick;
+		start_param.tmo_ev = ODP_EVENT_INVALID;
+
+		ret = odp_timer_restart(timer, &start_param);
+		CU_ASSERT_FATAL(ret == ODP_TIMER_SUCCESS);
+	}
+
+	ev = ODP_EVENT_INVALID;
+
+	if (cancel) {
+		ret = odp_timer_cancel(timer, &ev);
+		CU_ASSERT(ret == 0);
+
+		if (ret == 0)
+			CU_ASSERT(ev != ODP_EVENT_INVALID);
+	} else {
+		uint64_t diff_ns;
+
+		t1 = odp_time_global();
+		ev = wait_event(queue_type, queue, t1, 10 * tmo_ns);
+		t2 = odp_time_global();
+		diff_ns = odp_time_diff_ns(t2, t1);
+
+		CU_ASSERT(ev != ODP_EVENT_INVALID);
+		CU_ASSERT(diff_ns < 2 * tmo_ns);
+		CU_ASSERT((double)diff_ns > 0.5 * tmo_ns);
+	}
+
+	if (ev != ODP_EVENT_INVALID) {
+		CU_ASSERT(odp_event_type(ev) == ODP_EVENT_TIMEOUT);
+		tmo = odp_timeout_from_event(ev);
+		CU_ASSERT(odp_timeout_user_ptr(tmo) == USER_PTR);
+		CU_ASSERT(odp_timeout_timer(tmo) == timer);
```

Comment:
Things will go wrong if  event type != ODP_EVENT_TIMEOUT.

Reply:
Assert changed to fatal in v6.